confmap/provider/internal/configurablehttpprovider/provider.go

```diff
@@ -0,0 +1,147 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configurablehttpprovider // import "go.opentelemetry.io/collector/confmap/provider/internal/configurablehttpprovider"
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/confmap/provider/internal"
+)
+
+type TransportType string
+
+const (
+	PlainText TransportType = "http"
+	TLS       TransportType = "https"
+)
+
+type provider struct {
+	transport          TransportType
+	caCertPath         string // Used for tests
+	insecureSkipVerify bool   // Used for tests
+}
+
+// New returns a new provider that reads the configuration from http server using the configured transport mechanism.
+// There are two types of transport supported: PlainText (HTTP) and TLS (HTTPS).
+//
+// This internal provider supports "http" and "https" schemes based on the transport chosen, and can be called with a "uri" that
+// follows:
+//
+// One example for http-uri: http://localhost:3333/getConfig
+// One example for https-uri: https://localhost:3333/getConfig
+// This is used by the http and https external implementations.
+func New(transport TransportType) confmap.Provider {
+	return newConfigurableHTTPProvider(transport)
+}
+
+func newConfigurableHTTPProvider(transport TransportType) *provider {
+	return &provider{transport: transport, caCertPath: "", insecureSkipVerify: false}
+}
+
+func (fmp *provider) getHTTPClient() (*http.Client, error) {
+	return &http.Client{}, nil
+
+}
+
+func (fmp *provider) getHTTPSClient() (*http.Client, error) {
+	pool, err := x509.SystemCertPool()
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to create a cert pool: %w", err)
+	}
+
+	if fmp.caCertPath != "" {
+		cert, err := os.ReadFile(filepath.Clean(fmp.caCertPath))
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to read CA from %q URI: %w", fmp.caCertPath, err)
+		}
+
+		if ok := pool.AppendCertsFromPEM(cert); !ok {
+			return nil, fmt.Errorf("unable to add CA from uri: %s into the cert pool", fmp.caCertPath)
+		}
+	}
+
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: fmp.insecureSkipVerify,
+				RootCAs:            pool,
+			},
+		},
+	}, nil
+}
+
+// Get the client based on the type of transport that was selected.
+func (fmp *provider) getClient() (*http.Client, error) {
```

Comment:
This func can be merged with getHTTPClient and getHTTPSClient. I think will actually get cleaner than having 2 very small funcs and a large one.

Reply:
For reference of how it would look like:
```go
// Get the client based on the type of scheme that was selected.
func (fmp *provider) getClient() (*http.Client, error) {
	switch fmp.scheme {
	case HTTPScheme:
		return &http.Client{}, nil
	case HTTPSScheme:
		pool, err := x509.SystemCertPool()

		if err != nil {
			return nil, fmt.Errorf("unable to create a cert pool: %w", err)
		}
	
		if fmp.caCertPath != "" {
			cert, err := os.ReadFile(filepath.Clean(fmp.caCertPath))
	
			if err != nil {
				return nil, fmt.Errorf("unable to read CA from %q URI: %w", fmp.caCertPath, err)
			}
	
			if ok := pool.AppendCertsFromPEM(cert); !ok {
				return nil, fmt.Errorf("unable to add CA from uri: %s into the cert pool", fmp.caCertPath)
			}
		}
	
		return &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{
					InsecureSkipVerify: fmp.insecureSkipVerify,
					RootCAs:            pool,
				},
			},
		}, nil
	default:
		return nil, fmt.Errorf("invalid scheme type: %s", fmp.scheme)
	}
}
```

I prefer to be more verbose and having the 2 extra functions because we are encapsulating the logic of creation based on  schema type. This makes the code easier to understand and maintain IMHO: If we need to add more features in the https client creation, we will only change that function. we will not have the overhead of looking into the http one and the switch statement.

On a side note, I think I will rename the functions from `getXxxx` to `createXxxx` because we are effective creating something.

Reply:
I disagree with this, jumping from a 2lines func to a 2 lines func is not readable at all. Also the create funcs don't need to be on the struct but can be standalone funcs.

Reply:
Is this a merge blocker? If yes, I will address.

Reply:
We're discussing non-exported methods on a non-exported type in an internal package.  Can we land this PR and deal with the color of the bikeshed later?

Reply:
@Aneurysm9 usually when I create a PR and is about bikeshed, the maintainer is always right to get my work merged. So I am not sure why debating this :)

Reply:
I don't know why we're debating this either.  It doesn't matter.  You and I both know it doesn't matter.  Why put more effort into something that doesn't matter?  This has multiple approvals as-is.  Please merge it and create issues for any nits you want to pick.  They will be addressed.

Reply:
@bogdandrutu I put your suggestion in place in the last commit.